src/OrbitQt/FilterPanelWidget.h

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_QT_FILTER_PANEL_WIDGET_H_
+#define ORBIT_QT_FILTER_PANEL_WIDGET_H_
+
+#include <QFrame>
+#include <QWidget>
+
+#include "ui_FilterPanelWidget.h"
+
+namespace Ui {
+class FilterPanelWidget;
+}
+
+class FilterPanelWidget : public QFrame {
+  Q_OBJECT
+
+ public:
+  explicit FilterPanelWidget(QWidget* parent = nullptr);
+
+  void SetFilterFunctionsTextChangedCallback(
+      const std::function<void(const QString& text)>& callback) {
+    filter_functions_text_changed_callback_ = callback;
+  }
+
+  void SetFilterTracksTextChangedCallback(
+      const std::function<void(const QString& text)>& callback) {
+    filter_tracks_text_changed_callback_ = callback;
+  }
+
+  void SetFilterFunctionsText(const QString& text);
+  void SetTimerLabelText(const QString& text);
+
+ private slots:
+  void on_filterFunctions_textChanged(const QString& text);
+  void on_filterTracks_textChanged(const QString& text);
+
+ private:
+  Ui::FilterPanelWidget* ui;
+  std::function<void(const QString& text)> filter_functions_text_changed_callback_;
+  std::function<void(const QString& text)> filter_tracks_text_changed_callback_;
```

Comment:
It would probably be easier and more consistent with QT to not handle these with callbacks, but with Signals / Slots as well. You can introduce a public signal `FilterFunctionsTextChanged` in `FilterPanelWidget` and in `FilterPanelWidgetAction`. You should then be able to connect signal to signal and reduce the boilerplate quite a lot, i.e.:

In `FilterPanelWidget`:
```
QObject::connect(ui->filterFunctions, SIGNAL(TextChanged(QString)), this, SIGNAL(FilterFunctionsTextChanged(QString)));
```

And in `FilterPanelWidgetAction`:
```
QObject::connect(filter_panel_, SIGNAL(FilterFunctionsTextChanged(QString)), this, SIGNAL(FilterFunctionsTextChanged(QString)));
```

And then connect `OrbitMainWindow::OnFilterFunctionsTextChanged` to `FilterPanelWidgetAction::FilterFunctionsTextChanged`.

I've written this from the top of my head, so not 100% sure if the syntax is correct, but this should work.

Reply:
Done. Thanks so much! Using signals & slots simplifies the whole process a lot!